### ISSUE TITLE
Getting different results on different OSes

### ISSUE THREAD

**Issue by shadow-cs** [stryker-mutator/stryker-net#122]
Stryker is reporting different results on different OSes.

## TL;DR

While testing Stryker I stumbled upon an intriguing issue, Stryker was reporting some non-killed mutants which I didn't think twice about as it seemed it was working as expected. Then I uploaded the project to a private GitLab server and looked at CI results, one mutant was killed (which didn't happen on my Windows 10 PC).

## The issue

When running on Windows 10 ([log-20180928-windows-10.txt](https://github.com/stryker-mutator/stryker-net/files/2428754/log-20180928-windows-10.txt)) or Ubuntu under Windows 10 ([log-20180928-windows-10-linux.txt](https://github.com/stryker-mutator/stryker-net/files/2428755/log-20180928-windows-10-linux.txt)) all running .Net Core SDK 2.1.402 following mutant is marked as `Survived` but it shouldn't survive given the test case (see [MutationTesting.zip](https://github.com/stryker-mutator/stryker-net/files/2428761/MutationTesting.zip)).

```
Binary expression mutation on line 6: 'temp < 1000' ==> 'temp > 1000'
```

While running on Docker (image: `microsoft/dotnet:2.1.402-sdk`) it is killed as expected ([log-20180928-linux-docker.txt](https://github.com/stryker-mutator/stryker-net/files/2428756/log-20180928-linux-docker.txt)).

I was trying to replicate on public GitLab server, but I got this error:

```
[13:49:31 ERR] An error occurred during the mutationtest run 
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: MaxDegreeOfParallelism
   at System.Threading.Tasks.ParallelOptions.set_MaxDegreeOfParallelism(Int32 value)
   at Stryker.Core.MutationTest.MutationTestProcess.Test()
   at Stryker.Core.StrykerRunner.RunMutationTest(StrykerOptions options)
```

During my testing I also got both mutants killed on Windows 10, which happened after I deleted my native language folder from SDK installation so I could send a proper (English speaking) report. Cleaning output folders did fix this issue and results got back to "normal".

**Comment by shadow-cs** [comment]
If I run the test manually (see below), the test fails as expected.

```batch
set ActiveMutation=0
dotnet test --no-build --no-restore
```

When inspecting the processes spawned by Stryker, I always see `ActiveMutation` being `1`.

**Comment by Valdas3** [comment]
This seems to be related, so I will write here instead of creating a new issue.
I am trying to run same project ([StrykerTest.zip](https://github.com/stryker-mutator/stryker-net/files/2435583/StrykerTest.zip)) with some random tests multiple times and almost every time I run it I get different outcomes ([logs.zip](https://github.com/stryker-mutator/stryker-net/files/2435582/logs.zip)).
I am running it on windows 10 pc with 8 logical processors with dotnet core 2.1.300. My guess is that the issue might be with Test method of MutationTestProcess class: https://github.com/stryker-mutator/stryker-net/blob/6be7fe6c26a7c6ddeb472d5146e16cfb9fa629f7/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs#L97

Inside of it there is a Parallel.ForEach which (I think) tests the mutations. I tried setting usableProcessorCount to be equal to 1 and I was able to reproduce the expected results every time I run the project. So, my guess that there are some issues with thread safety, but I did not investigate further.

**Comment by simondel** [comment]
Thanks for testing this! It does seem like our sandboxes are not isolated properly.

**Comment by shadow-cs** [comment]
> Inside of it there is a Parallel.ForEach ...

I actually came to the same conclusion as @Valdas3 did. Could there be some race condition on the iterator in the `Parallel.ForEach`? Maybe simple `ToArray` might fix the issue. But I could be wrong on this in which case this could be an issue with the environment being somehow shared by multiple mutation processes.

I also think that the line:

https://github.com/stryker-mutator/stryker-net/blob/6be7fe6c26a7c6ddeb472d5146e16cfb9fa629f7/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs#L100

Should be changed to:

```csharp
var usableProcessorCount = Math.Max(logicalProsessorCount / 2, 1); 
```

Which should fix the `ArgumentOutOfRangeException`.

**Comment by richardwerkman** [comment]
@shadow-cs 
``` csharp
var usableProcessorCount = Math.Max(logicalProsessorCount / 2, 1); 
```

looks good, maybe we could add this to #111? Or is this important enough to create a new item for it?

**Comment by richardwerkman** [comment]
`var usableProcessorCount = Math.Max(logicalProsessorCount / 2, 1); ` will be fixed by #133 

But maybe there should be more research towards the thread safety.

**Comment by topennanen** [comment]
Looks like all the mutations use the same instance of executor, and therefore also the same instance of testrunner:
https://github.com/stryker-mutator/stryker-net/blob/b55988e12c1f9e6ef5ed09d969af862660b50d89/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestProcess.cs#L111 

Here the state of the testrunner gets changed, and we get a race condition. This also causes the varying results @Valdas3 reported.
https://github.com/stryker-mutator/stryker-net/blob/b55988e12c1f9e6ef5ed09d969af862660b50d89/src/Stryker.Core/Stryker.Core/MutationTest/MutationTestExecutor.cs#L32

Changing the runner calls to implicitly pass the mutation id fixed this.

**Comment by richardwerkman** [comment]
Great job on the research! It seems that the use of the same instance for every paralel testrun could be the problem. 

The method `SetActiveMutation()` sets a property inside the testrunner instance. This property is used during the `RunAll()` method call to start an sub-process with the id from the property in the environment of the sub-process. 

If the property changed in between the calls `SetActiveMutation()` and `RunAll()` the sub-process will use the wrong id in it's environment. 

Changing 
``` csharp
    public interface ITestRunner
    {
        TestRunResult RunAll(int? timeoutMS);
        void SetActiveMutation(int? id);
    }
```
to 
``` csharp
    public interface ITestRunner
    {
        TestRunResult RunAll(int? timeoutMS, int? activeMutationId);
    }
```
would work i guess. I'm not sure why I divided the setting of the active mutation and the testrun in the first place.

**Comment by shadow-cs** [comment]
> looks good, maybe we could add this to #111? Or is this important enough to create a new item for it?

Not at all that important. I was thinking about creating a PR once I had some time to spare.

One suggestion though, wouldn't it be better to create separate runners for each run/thread? This might prevent similar race-conditions in the future.

**Comment by simondel** [comment]
I think we should be able to reuse a process/thread to run multiple times. But parallel runs should be separated. So if we can keep the process/thread after a run to do another run, that would be great.

**Comment by Mobrockers** [comment]
@simondel @richardwerkman @shadow-cs Can you guys confirm this issue was fixed with #135 so this issue can be closed?

**Comment by shadow-cs** [comment]
@Mobrockers looks good, both runner isolation and `logicalProsessorCount` issue.

**Comment by maximal096** [comment]
This seems to be a problem again. Although you can now workaround it by manually settings the concurrency level as I did.

```
[22:39:47 WRN] Using a concurrency of 1 which is more than recommended 0 for normal system operation. This might have an impact on performance
```

This is on one of the shared gitlab runners they provide for free. Those might very well report a processor count of 0.

As far as I can tell the concurrency value is set here
https://github.com/stryker-mutator/stryker-net/blob/285afbed4116cdda8c35dd68fc4d7dc4fa8b393b/src/Stryker.Core/Stryker.Core/Options/Inputs/ConcurrencyInput.cs

There is ```Math.Max``` there anymore.

**Comment by rouke-broersma** [comment]
If the os reports a fake processor count to dotnet there's nothing we can do about it.

**Comment by maximal096** [comment]
You can't about not being able to get the processorCount. But you can not auto detect a value that is obviously wrong and will crash (like you did before). And set it to 1.

Imo there should be a check if the Default is 0 then print a warning along the lines of "Logical processor count is not reported by the system. The recommended concurrency is processorCount / 2."

The other warning is also meaningless in this case. Because you cannot calculate the recommended amount of concurrency in this case. There is nothing meaningful to compare the value to if the user set the concurrency.

**Comment by rouke-broersma** [comment]
That is a good suggestion. Could you create a feature request for it?

**Comment by maximal096** [comment]
Can do after work. Hope I remember.
